datasource.go

```diff
@@ -54,18 +55,30 @@ func NewDatasource(c Driver) datasource.ServeOpts {
 
 // QueryData creates the Responses list and executes each query
 func (ds *sqldatasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
-	// create response struct
-	response := backend.NewQueryDataResponse()
+	var (
+		response = backend.NewQueryDataResponse()
+		mtx      = sync.Mutex{}
+		wg       = sync.WaitGroup{}
+	)
 
+	wg.Add(len(req.Queries))
 	// Execute each query and store the results by query RefID
 	for _, q := range req.Queries {
-		frames, err := ds.handleQuery(q)
-		response.Responses[q.RefID] = backend.DataResponse{
-			Frames: frames,
-			Error:  err,
-		}
+		go func(query backend.DataQuery, mtx *sync.Mutex, wg *sync.WaitGroup) {
+			frames, err := ds.handleQuery(query)
+
+			mtx.Lock()
+			response.Responses[query.RefID] = backend.DataResponse{
+				Frames: frames,
+				Error:  err,
+			}
+			mtx.Unlock()
```

Comment:
If you pushed your result and query id to a channel, you wouldn't need mutexes: https://betterprogramming.pub/hands-on-go-concurrency-the-producer-consumer-pattern-c42aab4e3bd2

If you prefer the mutexes, it might be better if this were abstracted to it's own type that contains the responses and mutex. 

```go
type thing struct {
    mutex
    responses
}

(t thing) set(response) {
    m.Lock()
    defer m.Unlock()
    //do the stuff
}
```
Coupling the mutex with the result also make it harder to accidentally break in the future.

Reply:
We're using a channel in Grafana for the sql data sources https://github.com/grafana/grafana/blob/4ee037f523b9343f2bf4fe028ed19580b2110061/pkg/tsdb/sqleng/sql_engine.go#L159

Reply:
I liked the mutexes into a function instead of channels for this.

Reply:
Eventually I'd like to write more tests but testing anything with sql types is a struggle.